python_src/src/lamp_py/performance_manager/gtfs_utils.py

```diff
@@ -49,3 +63,184 @@ def start_time_to_seconds(
+        timestamp_lookup[min_timestamp] = int(result[0]["timestamp"])
+
+    events_dataframe["fk_static_timestamp"] = timestamp_lookup[
+        min(timestamp_lookup.keys())
+    ]
+    # add "fk_static_timestamp" column to trip update dataframe
+    # loop is to handle batches vehicle position batches that are applicable to
+    # overlapping static gtfs data
+    for min_timestamp in sorted(timestamp_lookup.keys()):
+        timestamp_mask = events_dataframe[timestamp_column] >= min_timestamp
+        events_dataframe.loc[
+            timestamp_mask, "fk_static_timestamp"
+        ] = timestamp_lookup[min_timestamp]
+
+    process_logger.log_complete()
+
+    return events_dataframe
+
+
+def add_parent_station_column(
+    events_dataframe: pandas.DataFrame,
+    db_manager: DatabaseManager,
+) -> pandas.DataFrame:
+    """
+    adds "parent_station" column to dataframe
+
+    events_dataframe must have "fk_static_timestamp" and "stop_id" columns
+
+    if "parent_station" value does not exist for a specific "stop_id", then
+    "stop_id" is used as "parent_station"
+    """
+    process_logger = ProcessLogger(
+        "add_parent_station",
+        row_count=events_dataframe.shape[0],
+    )
+    process_logger.log_start()
+
+    # handle dataframe with no rows
+    if events_dataframe.shape[0] == 0:
+        events_dataframe["parent_station"] = None
+        process_logger.log_complete()
+        return events_dataframe
+
+    # unique list of "fk_static_timestamp" values for pulling parent stations
+    static_timestamps = [
+        int(timestamp)
+        for timestamp in events_dataframe["fk_static_timestamp"].unique()
+    ]
+
+    # pull parent station data for joining to events dataframe
+    parent_station_query = sa.select(
+        StaticStops.timestamp.label("fk_static_timestamp"),
+        StaticStops.stop_id,
+        StaticStops.parent_station,
+    ).where(StaticStops.timestamp.in_(static_timestamps))
+    parent_stations = db_manager.select_as_dataframe(parent_station_query)
+
+    # join parent stations to events on "stop_id" and gtfs static
+    # timestamp foreign key
+    events_dataframe = events_dataframe.merge(
+        parent_stations, how="left", on=["fk_static_timestamp", "stop_id"]
+    )
+    # is parent station is not provided, transfer "stop_id" value to
+    # "parent_station" column
+    events_dataframe["parent_station"] = numpy.where(
+        events_dataframe["parent_station"].isna(),
+        events_dataframe["stop_id"],
+        events_dataframe["parent_station"],
+    )
+
+    process_logger.log_complete()
+
+    return events_dataframe
+
+
+def remove_bus_records(
+    events_dataframe: pandas.DataFrame,
+    db_manager: DatabaseManager,
+) -> pandas.DataFrame:
+    """
+    remove all records from dataframe associated with bus trips
+
+    events_dataframe must have "fk_static_timestamp" and "route_id" columns
+
+    route_type == 3 from the routes table indicates bus service, drop all records
```

Comment:
Remove BUS records by matching on `route_type == 3`. Requires that `fk_static_timestamp` columns was already added.